lua/lspconfig/util.lua

```diff
@@ -405,4 +388,14 @@ function M.path.exists(filename)
   return stat and stat.type or false
 end
 
+--- @deprecated use `vim.fs.find('.hg', { path = startpath, upward = true })` instead
+function M.find_mercurial_ancestor(startpath)
+  return vim.fs.find('.hg', { path = startpath, upward = true })
+end
+
+--- @deprecated use `vim.fs.find('node_modules', { path = startpath, upward = true })` instead
+function M.find_node_modules_ancestor(startpath)
+  return vim.fs.find('node_modules', { path = startpath, upward = true })
```

Comment:
one thing to note. I remember there was an issue about util.find_node_modules_ancestor

```
mkdir -p /home/mike/test/a/node_modules
mkdir /home/mike/test/node_modules
```

run `:lua print( require'lspconfig'.util.find_node_modules_ancestor('/home/mike/test/a'))` the result shouldbe test not a

so does there  need `limit =math.huge` with a stop and return the last item in the result list ?  🤔

Reply:
Oh wow, so the first "root marker" we find might not be the actual root marker? That's fucked up lol. OK, will test it out, thanks for the heads up.

Reply:
>  I assume this is not correct behavior for ".git" though?

yes something like git submodule

Reply:
it should return `test` not `test/a` ...i mean parent is actually root in this situation. maybe something like

```lua
  local resulst = vim.fs.find('node_modules', { path = startpath, upward = true, limit =math.huge, stop = vim.env.HOME })
 if #result > 2 then
      -- if find more and last is $HOME return #result - 1 should be root
    return { result[#result - 1] }

```

Reply:
What do you mean by "should" though? Wouldn't we have gotten 5000 complaints already if any web related lsp wasn't working correctly?

Reply:
That sounds like a recipe for disaster to fix something that will affect so many at the same time. I think we should just focus on pure refactor/rewrite for this PR and tackle the fix in a future PR.

Reply:
Related https://github.com/neovim/nvim-lspconfig/issues/1908. Took me some time to find it. I closed the issue because the original author did not respond. The plan is to fix it during the refactoring. So when I saw this PR, I thought of this issue. Of course, it can also be fixed in another PR.  not a big problem because there are only a few similar issues but  solved at the workspaceFolder level. But it is a bit strange that some of them do not support worksapceFolder .  it's fine.

Reply:
yes np and thanks. just it appear in my mind

Reply:
I think that issue seems more about the documentation of `find_node_modules_ancestor` rather than any problems it has as a helper function related to `lspconfig`. I think the `find_node_modules_ancestor` behaves "correctly" in that it seems to do what most people/servers expect. I could be wrong, but my interpretation is that we don't need to "fix" anything in this case.

Reply:
> I think the find_node_modules_ancestor behaves

emm I think it should return test instead of test/a , servers that support workspaceFolder, if you open test first and then open the file in a, the server will be reused. There is no problem here. Most people do this. But if you open a first and then open the file in test, it will be messed up. However, this is an isolated case, but it is still a problem.

Reply:
But as you said, this has nothing to do with PR. We shouldn't focus on this. Just keep doing it. No one will do it anyway. 😮‍💨